can/interfaces/dg/dg.py

```diff
@@ -0,0 +1,254 @@
+        isFD = msgForm["GCprotocol"]["body"]["data"]["status"] == 48
+        return Message(timestamp=timestamp, arbitration_id=headerForm,
+                       data=data, is_extended_id=extId, is_fd=isFD)
+
+    # Send method
+    def send(self, msg, timeout=None):
+        msgForm = self.decode_msg(msg)
+        self.beacon.FT_DATA_TX(self.channel, msgForm)
+
+    # Scheduled transmit
+    def _send_periodic_internal(self, msg, period, duration=None):
+        msgForm = {}
+        msgForm["message_list"] = [self.decode_msg(msg[0])]
+        msgForm["message_list"][0]["tx_period"] = int(period * 1000000)
+        msgForm["message_list"][0]["period_in_microsec"] = True
+        if duration is None:
+            cycles = 4294967295
+        else:
+            cycles = int(duration / period)
+        reply = self.beacon.CMD_SCHED_TX(self.channel, msgForm, iterations=cycles)
+        return Scheduling(msg, period, duration, reply["schedule_id"],
+                          weakref.ref(self.beacon), self.channel)
+
+    # Receive method
+    def _recv_internal(self, timeout=None):
+        if timeout is None:
+            reply = self.beacon.FT_DATA_WAIT_FOR_RX(timeout=.5)
+            while reply is None:
+                reply = self.beacon.FT_DATA_WAIT_FOR_RX(timeout=.5)
+            return self._encode_msg(reply), True
+
+        reply = self.beacon.FT_DATA_WAIT_FOR_RX(timeout=timeout)
+        if reply is not None:
+            return self._encode_msg(reply), True
+        return None, True
+
+    # Filtering method
+    def _apply_filters(self, filters):
+        if filters is None:
+            reply = self.beacon.CMD_CARD_GET_FILTER_HANDLES(self.channel)
+            for item in reply["GCprotocol"]["body"]["data"]["filter_handles"]:
+                self.beacon.CMD_CARD_MODIFY_FILTER(self.channel,
+                                                   self.beacon.DELETE_FILTER, filter_handle=item)
+            self.beacon.CMD_CARD_SET_DEFAULT_FILTER(self.channel, self.beacon.DEFAULT_FILTER_PASS)
+            return
+        dataFil = {}
+        counter = 0
+        dataFil["flags"] = (self.beacon.FILTER_FLAG_PASS
+                            | self.beacon.FILTER_FLAG_ACTIVE
+                            | self.beacon.FILTER_FLAG_OR_BLOCKS
+                            | self.beacon.FILTER_FLAG_SAMPLING_INACTIVE)
+        dataFil["filter_blocks"] = []
+        for item in filters:
+            dataFil["filter_blocks"].append({})
+            dataFil["filter_blocks"][counter]["byte_offset"] = 0
+            dataFil["filter_blocks"][counter]["data_type"] = self.beacon.FILTER_DATA_TYPE_HEADER
+            dataFil["filter_blocks"][counter]["operator"] = self.beacon.BIT_FIELD_CHECK
+            dataFil["filter_blocks"][counter]["mask"] = self._int2list(item["can_mask"])
+            dataFil["filter_blocks"][counter]["pattern"] = self._int2list(item["can_id"])
+            counter += 1
```

Comment:
What is counter used for?

Reply:
The counter is used to add and index the proper filter in "filters"

Reply:
Oh sorry, stupid me :)